model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt

```diff
@@ -147,10 +149,18 @@ class RepositoriesManager(val client: LocalModelClient) {
             ?: throw IllegalStateException("No version found for branch '${branch.branchName}' in repository '${branch.repositoryId}'")
     }
 
+    private val deltaCache = LRUMap<Pair<String, String?>, Lazy<Map<String, String>>>(10)
     fun computeDelta(versionHash: String, baseVersionHash: String?): Map<String, String> {
-        val version = CLVersion(versionHash, client.storeCache)
-        val baseVersion = baseVersionHash?.let { CLVersion(it, client.storeCache) }
-        return version.computeDelta(baseVersion)
+        return runSynchronized(deltaCache) {
+            deltaCache.getOrPut(versionHash to baseVersionHash) {
+                // lazy { ... } allows to run the computation without locking deltaCache
```

Comment:
Can there be a race condition that the cache is invalidated once the computation actually starts?

Reply:
The input is immutable. It never gets invalid.